src/jit/emitxarch.cpp

```diff
@@ -2847,15 +2847,34 @@ regNumber emitter::emitInsBinary(instruction ins, emitAttr attr, GenTree* dst, G
     assert(!dst->isContained() ||
            dst->isContainedIndir() ||
            dst->isContainedLclField() ||
+           dst->isContainedLclVar() ||  // this is essentially a contained LclField with zero offset
            instrIs3opImul(ins)); // dst on these isn't really the dst
 
+#ifdef DEBUG
     // src can be anything but both src and dst cannot be addr modes
     // or at least cannot be contained addr modes
     if (dst->isContainedIndir())
+    {
         assert(!src->isContainedIndir());
+    }
+
+    if (dst->isContainedLclVar())
+    {
+        assert(!src->isContainedIndir());
+        assert(!src->isContainedLclField());
+        assert(!src->isContainedLclVar());
+    }
 
     if (src->isContainedLclField())
+    {
         assert(!dst->isContained());
+    }
+
+    if (src->isContainedLclVar())
+    {
+        assert(!dst->isContained());
+    }
+#endif
 
```

Comment:
This is not easy to read or understand. I think it could be simplified. If there isn't already a IsContainedMem() you should add it - and a contained lclVar or lclFld should be considered ContainedMem.
Something like this would be much clearer:
```
if (dst->isContained())
{
    assert(!src->isContainedMem());
}
```
That would seem to summarize it (since the other kinds of containment on the src would be OK in this case).